### ISSUE TITLE
Incorrect "variant is never used" lint

### ISSUE THREAD

**Issue by LukasKalbertodt** [rust-lang/rust#44083]
A "variant is never used" lint is incorrectly triggered, presumably due to optimizations on MIR. I know my code sample looks pretty stupid as no human would write this, probably. But this kind of code might be generated. Anyway, the lint is incorrect I'd say.

I tried this code:

```rust
enum Baz {
    Foo,
    Bar,
}

fn main() {
    let x = Baz::Foo;
    match x {
        Baz::Foo => {},
        Baz::Bar if true => {},
        _ => {}
    }
}
```

I expected to see this happen: no warning

Instead, this happened: 
```
warning: variant is never used: `Bar`
 --> src/main.rs:3:5
  |
3 |     Bar,
  |     ^^^
  |
  = note: #[warn(dead_code)] on by default
```

## Meta

Happens with
- 1.19 stable,
- 1.20 beta,
- 1.21 nightly

**Comment by rkruppe** [comment]
Just to be clear: `Bar` isn't considered unused because the `Bar` arm is `main` not reached. It's considered unused because the variant is never *constructed*. If the variant is never constructed, it doesn't make any difference whether matches are prepared to handle it (and besides, those matches will never be executed -- by analogy, we also consider functions unused if they're called *from other dead code only*).

**Comment by LukasKalbertodt** [comment]
@rkruppe Oh right. Then I guess this can be closed, since it's a dupe of #19140
